### PR TITLE
Detect when changes have been made to a gem repo since the last version bump, and group messages by team

### PR DESCRIPTION
**Pull request by PeterHattyar**
https://trello.com/c/LMd1m6rj/3415-update-gem-release-alert-app-to-detect-when-changes-have-been-made-to-a-gem-repo-since-the-last-version-bump

### CODE REVIEW

File: lib/version_checker.rb [hunk 1]
```diff
@@ -10,58 +10,50 @@ class HttpError < RuntimeError
...
 
-    unless discrepancies_found
-      puts "No discrepancies found!"
+      puts team
+      puts message
     end
   end
 
-  def fetch_all_govuk_gemspecs
+  def fetch_all_govuk_gems
```
Comment (AgaDufrat) [inline lib/version_checker.rb]:
```suggestion
  def all_govuk_gems
```
Names of ruby methods typically describe what they return.

File: lib/version_checker.rb [hunk 2]
```diff
@@ -10,58 +10,50 @@ class HttpError < RuntimeError
...
+        Dir.chdir(repo) do
+          `git diff --name-only #{tag}`.split("\n")
         end
       end
     end
   end
 
-  def clone_github_repo(name)
-    `git clone --recursive --depth 1 --shallow-submodules git@github.com:alphagov/#{name}.git > /dev/null 2>&1`
+  def path_is_built_into_gem?(path)
```
Comment (AgaDufrat) [inline lib/version_checker.rb]:
```suggestion
  def path_built_into_gem?(path)
```
https://github.com/rubocop/ruby-style-guide?tab=readme-ov-file#predicate-methods-prefix

File: lib/version_checker.rb [hunk 3]
```diff
@@ -10,58 +10,50 @@ class HttpError < RuntimeError
...
-    end
+    fetch_all_govuk_gems.group_by { |gem| gem["team"] }.each do |team, gems|
+      message = gems.filter_map do |gem|
+        repo_name = gem["app_name"]
+        rubygems_version = fetch_rubygems_version(repo_name)
+        if !rubygems_version.nil? &&
+          files_changed_since_tag(repo_name, "v" + rubygems_version).any? { |path| path_is_built_into_gem?(path) }
+          "  #{repo_name} has unreleased changes since v#{rubygems_version}"
+        end
+      end.join("\n")
```
Comment (AgaDufrat) [inline lib/version_checker.rb]: `#print_version_discrepancies` is responsible for a lot of things:
 - finding the gems which have unreleased changes
 - formatting the message
 - grouping by the team

We should probably move some of this logic to separate methods.

Some of the patterns in this class don’t follow the style guide – I’d suggest configuring https://github.com/alphagov/rubocop-govuk first.

File: lib/version_checker.rb [hunk 4]
```diff
@@ -10,58 +10,50 @@ class HttpError < RuntimeError
...
-    unless discrepancies_found
-      puts "No discrepancies found!"
+      puts team
+      puts message
     end
   end
 
-  def fetch_all_govuk_gemspecs
+  def fetch_all_govuk_gems
     uri = URI("https://docs.publishing.service.gov.uk/gems.json")
```
Comment (AgaDufrat) [inline lib/version_checker.rb]: We could make this a constant (but I'm conscious it might be out of scope of this PR)

**Review by AgaDufrat — COMMENTED** [review]
Great work Peter! This really simplifies things. 

I understand that this is an intermediate step to update the logic. There’s scope for improvements but I think some of them would be premature and might be best to wait to polish off the code (especially the `print_version_discrepancies method`) once it’s actually sending notifications. 

I left a couple of inline comments. We should definitely configure rubocop-govuk for this repos and address any style violations.

We could also add some tests for unhappy paths.

File: lib/version_checker.rb (hunk 1 above)
Comment (PeterHattyar) [reply lib/version_checker.rb]: Great spot, adjusted.

File: lib/version_checker.rb (hunk 2 above)
Comment (PeterHattyar) [reply lib/version_checker.rb]: Didn't know about this, thanks! Changed accordingly.

File: lib/version_checker.rb (hunk 4 above)
Comment (PeterHattyar) [reply lib/version_checker.rb]: Seen no issue with it, so it's constant now.

File: lib/version_checker.rb (hunk 3 above)
Comment (PeterHattyar) [reply lib/version_checker.rb]: This is the big one, and I suggest I tackle this as a separate PR on its own, maybe even two separate ones, one splitting out the logics by responsibilities, and the other main one is the rubocop-govuk implementation?

**Review by AgaDufrat — APPROVED** [review]
Makes sense to address the rest of the comments in follow up PRs.

You may want to also create a `GovukGem` class with methods such as `.all` and another one that would return grouped results. Example in release app: https://github.com/alphagov/release/blob/main/app/models/repo.rb

I’m also not sure what the error handling is trying to achieve. I think it will mask the unsuccessful Net::HTTP response, which will make it difficult to debug, so probably worth revisiting it and adding tests.

**Pull request merged**